Update the service resolver documentation to include cluster peering failover and redirects (#14513)
hashicorp/consul: website/content/docs/connect/config-entries/service-resolver.mdx

</CodeTabs>

> Enable failover to a WAN federated datacenter and then a cluster peer for all

Reviewer: I saw the edits you made for this. I like it! Had a few minor follow-up suggestions.

service subsets.

<CodeTabs tabs={[ "HCL", "Kubernetes YAML", "JSON" ]}>

```hcl
Kind           = "service-resolver"
Name           = "web"
ConnectTimeout = "15s"
Failover = {
  "*" = {
    Targets = [
      {Datacenter = "dc3"},
      {Peer = "peer-01"}
    ]
  }
}
```

```yaml
apiVersion: consul.hashicorp.com/v1alpha1
kind: ServiceResolver
metadata:
  name: web
spec:
  connectTimeout: 15s
  failover:
    '*':
      targets:
      - datacenter: "dc3"
      - peer: "peer-01"
```

```json
{
  "Kind": "service-resolver",
  "Name": "web",
  "ConnectTimeout": "15s",
  "Failover": {
    "*": {
      "Targets": [
        {"Datacenter": "dc3"},

>         {"Peer": "peer-01"}

Author: Nice catch

          description:
            'Specifies the destination datacenter to resolve the service from.',
        },
        {
          name: 'Peer',
          type: 'string: ""',
          description:
            `Specifies the destination cluster peer to resolve the service

>                 from. When peer is specified, the failover target relies on Envoy's
>                 outlier detection for health checks. Consul relies on outlier
>                 detection because imported service health checks don't account

Reviewer 2: ```suggestion
                from. When the peer is specified, the failover target relies on Envoy's
                outlier detection for health checks. Consul relies on outlier
                detection because imported service health checks do not account
```
No contractions per the style guide

          description:
            'Specifies the destination datacenter to resolve the service from.',
        },
        {
          name: 'Peer',
          type: 'string: ""',
          description:
              `Specifies the destination cluster peer to resolve the target service name from.
                  Ensure that [intentions are defined](/docs/connect/cluster-peering/create-manage-peering#authorize-services-for-peers)

>                   on the peered cluster to allow the source service to call the upstream service. When

Reviewer: @erichaberkorn : does this seem like a helpful clarification (that the upstream we're referring to is the service in this redirect clause)?

```suggestion
                  on the peered cluster to allow the source service to access this redirect target service as an upstream.
```

                string \`"*"\` is a wildcard that applies to any subset not otherwise
                specified here.
                <br><br>

>                 \`Service\`, \`ServiceSubset\`, \`Namespace\`, \`Targets\`, and
>                 \`Datacenters\` cannot all be empty at once.`,

Reviewer 3: ```suggestion
                At least one of \`Service\`, \`ServiceSubset\`, \`Namespace\`, \`Targets\`, or
                \`Datacenters\` must be specified.`,
```

        yaml: `Controls when and how to
                reroute traffic to an alternate pool of service instances.
                <br><br>
                The map is keyed by the service subset it applies to and the special
                string \`"*"\` is a wildcard that applies to any subset not otherwise
                specified here.
                <br><br>

>                 \`service\`, \`serviceSubset\`, \`namespace\`, \`targets\` and
>                 \`datacenters\` cannot all be empty at once.`,

Reviewer 3: ```suggestion
                At least one of \`service\`, \`serviceSubset\`, \`namespace\`, \`targets\` and
                \`datacenters\` must be specified.`,
```

                'The named subset to use for the failover target. If empty, the default subset for the requested service name is used.',
            },
            {
              name: 'Namespace',
              enterprise: true,
              type: 'string: ""',
              description:
                `The namespace to use for the failover target. If empty, the \`default\` namespace is used.`,
            },
            {
              name: 'Partition',
              enterprise: true,
              type: 'string: ""',
              description:
                `The admin partition within the same datacenter to use for the failover target.
                If empty, the \`default\` partition is used.
                To use an admin partition in a different datacenter for the failover target,
                use the \`Peer\` field instead.`,
            },
            {
              name: 'Datacenter',
              type: 'string: ""',
              description:
                `The WAN federated datacenter to use for the failover target.
                If empty, the current datacenter is used.
                To use a datacenter for the failover target that is connected

>                 with a cluster peering relationship rather than WAN federation ,

Reviewer: ```suggestion
                with a cluster peering relationship rather than WAN federation,
```

                use the \`Peer\` field instead.`,
            },
            {
              name: 'Peer',
              type: 'string: ""',
              description:
              `Specifies the destination cluster peer to resolve the target service name from.
                  Ensure that [intentions are defined](/docs/connect/cluster-peering/create-manage-peering#authorize-services-for-peers)

>                   on the peered cluster to allow the source service to call the upstream service.

Reviewer: @erichaberkorn : does this seem like a helpful clarification (that the upstream we're referring to is the service in this failover target).

```suggestion
                  on the peered cluster to allow the source service to access this failover target service as an upstream.
```

          type: 'array<string>',
          description: 'A fixed list of datacenters to try during failover.',
        },
        {
          name: 'Targets',
          type: 'array<ServiceResolverFailoverTarget>',
          description: `A fixed list of failover targets to try during
                          failover. It allows operators to express complicated failover
                          scenarios such as between admin partitions, datacenters and cluster
                          peers.`,
          children: [
            {
              name: 'Service',
              type: 'string: ""',
              description:
                'The service to use instead of the default of the failover target. If empty the current service is used.',
            },
            {
              name: 'ServiceSubset',
              type: 'string: ""',
              description:
                'The named subset to use for the failover target. If empty the default subset for the requested service is used.',
            },
            {
              name: 'Namespace',
              enterprise: true,
              type: 'string: ""',
              description:

>                 'The namespace to use for the failover target. If empty the default namespace is used.',

Author: I left out the comma for consistency with everything else. I'll make this change across the entire file

>                 'The namespace to use for the failover target. If empty the default namespace is used.',

Reviewer: Gotcha, thanks. I find that things like this can have a surprisingly large impact on readability when it's something you aren't familiar with.

            },
            {
              name: 'Partition',
              enterprise: true,
              type: 'string: ""',
              description:

>                 'The admin partition to use for the failover target. If empty the default partition is used.',

Reviewer: When we say "default partition", do we mean literally the `default` partition? Or it defaults to the partition of this service resolver?

Also, we should provide some guidance on when to use `Partition` vs `Peer`.
```suggestion
                `The admin partition within the same datacenter to use for the failover target.
                If empty, the default partition is used.
                To use an admin partition in a different datacenter for the failover target,
                use the \`Peer\` field instead.`,
```

            },
            {
              name: 'Datacenter',
              type: 'string: ""',
              description:

>                 'The datacenter to use for the failover target. If empty the current datacenter is used.',

Reviewer: Also, we should provide some guidance on when to use `Datacenter` vs `Peer`.

```suggestion
                `The WAN federated datacenter to use for the failover target.
                If empty, the current datacenter is used.
                To use a datacenter for the failover target that is connected
                with a cluster peering relationship rather than WAN federation ,
                use the \`Peer\` field instead.`,
```

            },
            {
              name: 'Peer',
              type: 'string: ""',
              description:

>               `Specifies the destination cluster peer to resolve the service
>                   from. When peer is specified, the failover target relies on Envoy's
>                   outlier detection for health checks. Consul relies on outlier
>                   detection because imported service health checks don't account
>                   for service routers, splitters and resolvers.`,

Reviewer 2: ```suggestion
              `Specifies the destination cluster peer to resolve the service
                  from. When the peer is specified, the failover target relies on Envoy's
                  outlier detection for health checks. Consul relies on outlier
                  detection because imported service health checks do not account
                  for service routers, splitters and resolvers.`,
```

>               `Specifies the destination cluster peer to resolve the service
>                   from. When the peer is specified, the failover target relies on Envoy's
>                   outlier detection for health checks. Consul relies on outlier
>                   detection because imported service health checks do not account
>                   for service routers, splitters and resolvers.`,

Reviewer: My attempt at clarifying:
- What Envoy outlier detection is for someone unfamiliar with the concept
- What an imported service health check is

```suggestion
              `Specifies the destination cluster peer to resolve the target service name from.
                  When the peer name is specified, Consul uses Envoy's outlier detection
                  to determine the health of the failover target based on
                  whether communication attempts to the failover target are generally successful.
                  Service health checks imported from a peer are not considered in the health
                  of a failover target because they do not account for service routers, splitters, and resolvers
                  that may be defined in the peer for the target service.`,
```
